src/bech32.c

```diff
@@ -138,9 +149,15 @@ static int bech32_decode(char *hrp, uint8_t *data, size_t *data_len, const char
         ++i;
     }
     if (have_lower && have_upper) {
-        return 0;
+        return BECH32_ENCODING_NONE;
+    }
+    if (chk == bech32_final_constant(BECH32_ENCODING_BECH32)) {
+        return BECH32_ENCODING_BECH32;
+    } else if (chk == bech32_final_constant(BECH32_ENCODING_BECH32M)) {
+        return BECH32_ENCODING_BECH32M;
+    } else {
+        return BECH32_ENCODING_NONE;
```

Comment:
return the checksum constant directly, or e.g. BECH32_CHECKSUM_UNKNOWN if invalid